### DISCUSSION TITLE
Adding markers as scatter plot data points over a displayed image (PlotItem methods inheritance)

### DISCUSSION THREAD

**Discussion by ghylander** [pyqtgraph/pyqtgraph#3030] [Q&A]
Hi all,

I have a fairly complete pyqtgraph app that I unfortunately cannot share, so please bear with me.
I'm trying to implement what can be read in the title.

The relevant bits of the app:

1. In the app, I have a window with 2 GraphicsView objects.
2. Inside each GraphicsView object, there is a GraphicsLayout object.
3. Inside each GraphicsLayout, there is a PlotItem object, created via the addPlot method.
4. Inside each PlotItem object, there is an ImageItem, added via the addItem method.

What I'm trying to achieve is a feature that allows me to left click on one of the images being displayed, and add a marker on that location.
I haven't seen this implemented in pyqtgraph, so I'm implementing it by adding a (initially empty) scatter plot over the image, then adding the location clicked as a data point in the scatter plot.

This behaviour would be toggled on or off via a button (click to enable marking). When enabled, this overrides the default mouseClickEvent. When disabled, the regular mouseClickEvent is in place.

For the implmentation of the mouseClickEvent override, I've been reading the docs, source and provided examples.
In particular, the "Crosshair / Mouse interaction" example seemed quite fitting as a starting point.

However, I'm having trouble tracing back the PlotItem object's 'scene' and 'sceneBoundingRect' methods.

These (specially the scene method) seem key in the process, but after tracing inheritance down to QtWidgets objects such as QGraphicsObject, I have not found the definition of this method.

I could probably just achieve my goal without this knowledge, but I really don't like working with objects and methods I don't fully understand.

Any help appreciated (including ideas regarding the marking of the images).

**Comment by j9ac9k** [comment]
You don't necessarily need to override mouseClickEvent, you can probably make your own slot that catches the `GraphicsScene.sigMouseClicked` signal, and the function/method there can check the status of of the "global toggle" for enabling/disabling.  Furthermore, the signal carries the event that contains the `.scenePos()` method which returns the position in scene coordinates.

Then, you can call `ViewBox.mapFromScene(event.scenePos())` and get coordinates in the view-space.

If you can post some basic code of showing a dummy image item and a scatter plot, I can give a jab at adding that bit of mouse interaction.

**Reply by ghylander** [reply]
Thank you for your reply.

Having gone in blind, below is what I had implemented to handle the mouse events using the example I cited as starting point (`scatterEdit` is a boolean whose state is toggled by a button).

I don't understand what you mean by creating a slot to cache the `sigMouseClicked` event. Care to elaborate?

On the other hand, I was just starting to implement a way to delete points from the scatter plot, but it seems there is no direct way, is there?
Is the only way to `getData`, remove the point, `clear` the plot, then `setData`?

```
image_view_ori = pg.GraphicsView()
image_layout_ori = pg.GraphicsLayout()
p1 = image_layout_ori.addPlot(title="Original")
p1.invertY()
p1.setAspectLocked()
image_view_ori.setCentralItem(image_layout_ori)
right_layout.addWidget(image_view_ori, 0, 0)
vb = p1.vb

def mouseClicked(ev):
    if ev.button() == QtCore.Qt.MouseButton.LeftButton and scatterEdit:
        mousePoint = vb.mapSceneToView(ev.scenePos())

        scatterPlot.addPoints([{"pos": (mousePoint.x(), mousePoint.y()), 
                                "pen": 'penVar,
                                "symbol": symbolVar,
                                }])


def pointRightClicked(self, points, ev):
    # if ev.button() == QtCore.Qt.MouseButton.RightButton and scatterEdit:
        print(ev)
        print(points)
        print(scatterPlot.__dir__())
        #Here I'm just trying to check if there is any built-in method to remove points from the plot


p1.scene().sigMouseClicked.connect(mouseClicked)


# Item for displaying image data
img_ori = pg.ImageItem()
scatterPlot = pg.ScatterPlotItem()
p1.addItem(img_ori)
p1.addItem(scatterPlot)
scatterPlot.sigClicked.connect(pointRightClicked)
```

**Reply by j9ac9k** [reply]
The following code adds a scatter plot circle where the mouse is clicked:

```python
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore

pg.mkQApp()

image_view_ori = pg.GraphicsView()
image_layout_ori = pg.GraphicsLayout()
p1 = image_layout_ori.addPlot(title="Original")
p1.invertY()
p1.setAspectLocked()
image_view_ori.setCentralItem(image_layout_ori)
# right_layout.addWidget(image_view_ori, 0, 0)
vb = p1.vb

def mouseClicked(ev):
    # if ev.button() == QtCore.Qt.MouseButton.LeftButton and scatterEdit:
    mousePoint = p1.vb.mapSceneToView(ev.scenePos())

    scatterPlot.addPoints(
        [
            {
                "pos": (mousePoint.x(), mousePoint.y()), 
                "pen": 'white',
                "symbol": 'o',
            }
        ]
    )


def pointRightClicked(self, points, ev):
    # if ev.button() == QtCore.Qt.MouseButton.RightButton and scatterEdit:
    print(ev)
    print(points)
    print(scatterPlot.__dir__())
    #Here I'm just trying to check if there is any built-in method to remove points from the plot


p1.scene().sigMouseClicked.connect(mouseClicked)

# Item for displaying image data
img_ori = pg.ImageItem()
scatterPlot = pg.ScatterPlotItem()
p1.addItem(img_ori)
p1.addItem(scatterPlot)
scatterPlot.sigClicked.connect(pointRightClicked)
image_view_ori.show()

if __name__ == "__main__":
    pg.exec()
```

**Reply by ghylander** [reply]
Hi,

I'm unsure of what's changed in your code with respect to mine?

The `marker` and `pen` arguments of the point added to the scatter plot are defined by the placeholder `penVar` and `markerVar` that I left in the pseudo-code.
Is this the same approach as you described in your original comment?

**Reply by j9ac9k** [reply]
Oh, the approach is the same, the above code is a standalone example which runs, that's all :)

Is there behavior you're trying to get that this doesn't demonstrate? to get the key toggle on/off bit, you would need to create a subclass of QWidget (or maybe GraphicsView) and re-implement your own [`keyPressEvent`](https://doc.qt.io/qt-6/qwidget.html#keyPressEvent) method that "listens" for key-presses.  

Here's an example that captures the toggling of the listening of mouse events, you could do something fancier with tying the key-press to the disconnecting of the signal ...

```python
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore, QtGui


pg.mkQApp()

listenToClicks = False

class MyGraphicsView(pg.GraphicsView):
    mouseClickReceiveToggled = QtCore.Signal()

    def keyPressEvent(self, event: QtCore.QEvent) -> bool:
        if event.type() == QtCore.QEvent.Type.KeyPress:
            if event.key() == QtCore.Qt.Key.Key_B:
                self.mouseClickReceiveToggled.emit()
        return super().keyPressEvent(event)


image_view_ori = MyGraphicsView()
image_layout_ori = pg.GraphicsLayout()
p1 = image_layout_ori.addPlot(title="Original")
p1.invertY()
p1.setAspectLocked()
image_view_ori.setCentralItem(image_layout_ori)
# right_layout.addWidget(image_view_ori, 0, 0)
vb = p1.vb

def toggleMouseClickListen():
    global listenToClicks
    listenToClicks = not listenToClicks
    print(f"Mouse Clicks Processed: {listenToClicks}")


def addScatterPoint(point):
    global listenToClicks
    if not listenToClicks:
        return None
    scatterPlot.addPoints(
        [
            {
                "pos": (point.x(), point.y()), 
                "pen": 'white',
                "symbol": 'o',
            }
        ]
    )

def mouseClicked(ev):
    # if ev.button() == QtCore.Qt.MouseButton.LeftButton and scatterEdit:
    mousePoint = p1.vb.mapSceneToView(ev.scenePos())
    addScatterPoint(mousePoint)

image_view_ori.mouseClickReceiveToggled.connect(toggleMouseClickListen)
p1.scene().sigMouseClicked.connect(mouseClicked)

# Item for displaying image data
img_ori = pg.ImageItem()
scatterPlot = pg.ScatterPlotItem()
p1.addItem(img_ori)
p1.addItem(scatterPlot)
# scatterPlot.sigClicked.connect(pointRightClicked)
image_view_ori.show()

if __name__ == "__main__":
    pg.exec()
```

**Reply by ghylander** [reply]
Hi,

No, the behaviour works as intended.
I was surprised because, in your first reply, you seemed to mention a different approach.

For the toggle, at first I tried to use a `QPushButton` and connect a function to the button's `clicked()` signal.
However, because I needed the button to be exclusive with a different button, and had some code that was checking whether the button was active or not, I ended up using the `QPushButton.setCheckable()` and `QPushButton.setAutoExclusive()` methods, then just check whether a button is `isChecked()`.

Thanks for the heads up about capturing key presses or combinations to toggle the buttons, I can link that to the the `QPushButton.toggle()` slot.

To remove data points from the scatter plot I did end up simply connecting the `scatterPlot.sigClicked()` signal to a function, then using `ScatterPlotItem.getData()`, and deleting the point by index, finally using `ScatterPlotItem.setData()`. Not ideal, but works for my use case as I'm not expecting large amount of data points in the scatter plot.

**Reply by j9ac9k** [reply]
If you're using a `QPushButton` you can specify a keyboard shortcut in its constructor that will trigger the toggled signal, so you don't even have to subclass and listen for the `QKeyEvent`.  See here: https://doc.qt.io/qt-6/qpushbutton.html#details

The mechanism I described was using the `sigMouseClicked` signal from `GraphicsScene` and converting it using `ViewBox.mapSceneToView(event.scenePos())` (the `.mapSceneToView` method probably exists in the higher level plot object too).  I posted my code as it was unclear if you didn't have something working as intended or not.

For removing the points, that's probably how I would go about it as well, I agree, not ideal... I'd be open to a PR, that adds an optional `index` argument to the `ScatterPlotItem.clear()` allowing for the removal of a subset of points by index.
